cql3/expr/expression.cc

```diff
@@ -343,43 +343,37 @@ bool_or_null limits(const expression& lhs, oper_t op, const expression& rhs, con
-    auto&& element_type = col_type->is_set() ? col_type->name_comparator() : col_type->value_comparator();
-    return value.with_linearized([&] (bytes_view val) {
-        auto exists_in = [&](auto&& range) {
-            auto found = std::find_if(range.begin(), range.end(), [&] (auto&& element) {
-                return element_type->compare(element.serialize_nonnull(), val) == 0;
-            });
-            return found != range.end();
-        };
-        if (col_type->is_list()) {
-            return exists_in(value_cast<list_type_impl::native_type>(collection));
-        } else if (col_type->is_set()) {
-            return exists_in(value_cast<set_type_impl::native_type>(collection));
-        } else if (col_type->is_map()) {
-            auto data_map = value_cast<map_type_impl::native_type>(collection);
-            using entry = std::pair<data_value, data_value>;
-            return exists_in(data_map | transformed([] (const entry& e) { return e.second; }));
-        } else {
-            throw std::logic_error("unsupported collection type in a CONTAINS expression");
-        }
-    });
-}
 
-/// True iff a column is a collection containing value.
-bool contains(const expression& collection_val, const raw_value_view& value, const evaluation_inputs& inputs) {
-    const data_type collection_type = type_of(collection_val);
-    const managed_bytes_opt collection_bytes = evaluate(collection_val, inputs).to_managed_bytes_opt();
-    if (collection_bytes) {
-        return contains(collection_type->deserialize(managed_bytes_view(*collection_bytes)), value);
+    const abstract_type& lhs_type = type_of(lhs)->without_reversed();
+    data_value lhs_collection = lhs_type.deserialize(managed_bytes_view(sides_bytes->first));
+
+    const collection_type_impl* collection_type = dynamic_cast<const collection_type_impl*>(&lhs_type);
+    data_type element_type =
+        collection_type->is_set() ? collection_type->name_comparator() : collection_type->value_comparator();
+
+    auto exists_in = [&](auto&& range) {
+        auto found = std::find_if(range.begin(), range.end(), [&](auto&& element) {
+            return element_type->compare(managed_bytes_view(element.serialize_nonnull()), sides_bytes->second) == 0;
+        });
+        return found != range.end();
+    };
+    if (collection_type->is_list()) {
+        return exists_in(value_cast<list_type_impl::native_type>(lhs_collection));
+    } else if (collection_type->is_set()) {
+        return exists_in(value_cast<set_type_impl::native_type>(lhs_collection));
+    } else if (collection_type->is_map()) {
+        auto data_map = value_cast<map_type_impl::native_type>(lhs_collection);
+        using entry = std::pair<data_value, data_value>;
+        return exists_in(data_map | transformed([](const entry& e) { return e.second; }));
     } else {
-        return false;
+        on_internal_error(expr_logger, "unsupported collection type in a CONTAINS expression");
```

Comment:
I hope we have an earlier check that lhs is indeed a collection, and a malicious request cannot trigger this case?

Reply:
`prepare_expression` should perform all necessary type checks. `evaluate()` usually assumes that the types are correct, but there are some type checks sprinkled here and there.

Reply:
`prepare_expression` should perform all type checks. `evaluate()` usually assumes that the types were checked and are correct, although there are some type checks sprinkled here and there.